x/pairing/keeper/unresponsive_provider.go

```diff
@@ -0,0 +1,314 @@
+		if punishEntry.providerServicedCu == 0 {
+			chainID := strings.Split(punishEntryKey, "_")[1]
+
+			// Try getting the provider's stake entry
+			_, foundStakeEntry := k.epochStorageKeeper.GetEpochStakeEntries(ctx, currentEpoch, epochstoragetypes.ProviderKey, chainID)
+
+			// if stake entry is not found -> this is the provider's first epoch (didn't get a chance to stake yet) -> don't punish him
+			if !foundStakeEntry {
+				continue
+			}
+		}
+
+		// The provider served less CU than the sum of the complainers CU -> should be punished
+		if punishEntry.complainersUsedCu > punishEntry.providerServicedCu {
+			providersToPunishMap[punishEntryKey] = punishEntry.providerPaymentStorageIndexList
+		}
+	}
+
+	return providersToPunishMap, nil
+}
+
+// Function to count the CU serviced by the unresponsive provider or the CU of the complainers.
+// The function counts CU from <epoch> back. The number of epochs to go back is <epochsNumToGoBack>. The function returns the last visited epoch
+func (k Keeper) countCuForUnresponsiveProviderAndComplainers(ctx sdk.Context, isProvider bool, epoch uint64, epochsNumToGoBack uint64, providersPunishEntryMap map[string]punishEntry) (uint64, error) {
+	epochTemp := epoch
+
+	for counter := uint64(0); counter < epochsNumToGoBack; counter++ {
+		// Get previous epoch (from epochTemp)
+		previousEpoch, err := k.epochStorageKeeper.GetPreviousEpochStartForBlock(ctx, epochTemp)
+		if err != nil {
+			return 0, utils.LavaError(ctx, k.Logger(ctx), "get_previous_epoch", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", epochTemp)}, "couldn't get previous epoch")
+		}
+
+		// Get the current stake storages (from all chains)
+		stakeStorageList := k.getCurrentProviderStakeStorageList(ctx)
+		if len(stakeStorageList) == 0 {
+			// no provider is staked -> CU = 0
+			return 0, nil
+		}
+
+		// Get a providerPaymentStorage of all staked providers
+		providerPaymentStorageList, err := k.getProviderPaymentStorageFromStakeStorageList(ctx, previousEpoch, stakeStorageList)
+		if err != nil {
+			return 0, utils.LavaError(ctx, k.Logger(ctx), "get_provider_payment_storage_from_stake_storage_list", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", previousEpoch)}, "couldn't get providerPaymentStorageList from stakeStorageList")
+		}
+
+		// Go over providerPaymentStorage list
+		for _, providerPaymentStorage := range providerPaymentStorageList {
+			cuCount := uint64(0)
+
+			// Get provider address and chainID
+			providerAddress := k.getProviderAddressFromProviderPaymentStorageKey(providerPaymentStorage.Index)
+			chainID := k.getChainIDFromProviderPaymentStorageKey(providerPaymentStorage.Index)
+
+			// Create a punishEntryKey from the provider's address and the chain ID
+			punishEntryKey := providerAddress + "_" + chainID
+
+			// find the provider's punishEntry in map
+			providerPunishEntry, punishEntryFound := providersPunishEntryMap[punishEntryKey]
+
+			if isProvider {
```

Comment:
no need for calling this function twice with this boolean, create a function for counting providers and counting complainers and do both

Reply:
done